broker/src/main/java/io/camunda/zeebe/broker/transport/partitionapi/InterPartitionCommandSenderImpl.java

```diff
@@ -0,0 +1,98 @@
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.transport.partitionapi;
+
+import io.atomix.cluster.MemberId;
+import io.atomix.cluster.messaging.ClusterCommunicationService;
+import io.camunda.zeebe.broker.Loggers;
+import io.camunda.zeebe.broker.partitioning.topology.TopologyPartitionListenerImpl;
+import io.camunda.zeebe.clustering.management.InterPartitionMessageEncoder;
+import io.camunda.zeebe.clustering.management.MessageHeaderEncoder;
+import io.camunda.zeebe.engine.transport.InterPartitionCommandSender;
+import io.camunda.zeebe.protocol.record.ValueType;
+import io.camunda.zeebe.protocol.record.intent.Intent;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import org.agrona.ExpandableArrayBuffer;
+import org.agrona.MutableDirectBuffer;
+import org.slf4j.Logger;
+
+public final class InterPartitionCommandSenderImpl implements InterPartitionCommandSender {
+  private static final Logger LOG = Loggers.TRANSPORT_LOGGER;
+  private final ClusterCommunicationService communicationService;
+
+  private final TopologyPartitionListenerImpl partitionListener;
+  private final Encoder encoder = new Encoder();
+
+  public InterPartitionCommandSenderImpl(
+      final ClusterCommunicationService communicationService,
+      final TopologyPartitionListenerImpl partitionListener) {
+    this.communicationService = communicationService;
+    this.partitionListener = partitionListener;
+  }
+
+  @Override
+  public void sendCommand(
+      final int receiverPartitionId,
+      final ValueType valueType,
+      final Intent intent,
+      final BufferWriter command) {
+    final var partitionLeaders = partitionListener.getPartitionLeaders();
+    if (!partitionLeaders.containsKey(receiverPartitionId)) {
+      LOG.warn(
+          "Not sending command {} {} to {}, no known leader for this partition",
+          intent,
+          valueType,
+          receiverPartitionId);
+      return;
+    }
+    final int partitionLeader = partitionLeaders.get(receiverPartitionId);
+
+    LOG.trace(
+        "Sending command {} {} to partition {}, leader {}",
+        valueType,
+        intent,
+        receiverPartitionId,
+        partitionLeader);
+
+    final var message = encoder.encode(receiverPartitionId, valueType, intent, command);
+
+    communicationService.unicast(
+        TOPIC_PREFIX + receiverPartitionId, message, MemberId.from("" + partitionLeader));
+  }
+
+  private static final class Encoder {
+    private final MessageHeaderEncoder headerEncoder = new MessageHeaderEncoder();
+    private final InterPartitionMessageEncoder bodyEncoder = new InterPartitionMessageEncoder();
+    private final MutableDirectBuffer messageBuffer = new ExpandableArrayBuffer();
+    private final MutableDirectBuffer commandBuffer = new ExpandableArrayBuffer();
```

Comment:
@deepthidevaki I'm still using expandable buffers here but copying the correct amount of bytes before returning. I've found no way to not allocate a new byte array here because command is of variable length.

Reply:
Okay, I'm doing that now :+1: 
But regarding thread safety, the header and body encoders are still unsafe to use concurrently so I'd have to create them ad-hoc as local variables too, right? Do you think that's worth it?

Reply:
Yes. I think so. We know that that the sender is always called from the same thread. But it is safer not to make that assumption. Also there is no disadvantage in creating adhoc, right?

Reply:
> Also there is no disadvantage in creating adhoc, right?

Nothing apart from allocating but looking at the encoders, that should be fairly cheap anyway :+1: